Sources/ScipioKit/Producer/FrameworkProducer.swift

```diff
@@ -0,0 +1,136 @@
+        fileSystem: any FileSystem = localFileSystem
+    ) {
+        self.rootPackage = rootPackage
+        self.buildOptions = buildOptions
+        self.cacheMode = cacheMode
+        self.fileSystem = fileSystem
+    }
+
+    func produce(mode: Runner.Mode, outputDir: URL) async throws {
+        let cacheSystem = CacheSystem(rootPackage: rootPackage,
+                                      buildOptions: buildOptions,
+                                      outputDirectory: outputDir,
+                                      storage: cacheStorage)
+        let compiler = Compiler(rootPackage: rootPackage, buildOptions: buildOptions)
+        try await compiler.clean()
+
+        let products = productsToBuild(for: mode)
+        for product in products {
+            if product.isBinaryTarget {
+                logger.error("🚧 \(product.target) is binaryTarget. Currently, binaryTarget is not supported. Skip it.")
+                continue
+            } else {
+                try await buildXCFrameworkIfNeeded(
+                    product,
+                    mode: mode,
+                    outputDir: outputDir,
+                    cacheSystem: cacheSystem,
+                    compiler: compiler
+                )
+            }
+
+            if isCacheEnabled {
+                let outputPath = outputDir.appendingPathComponent(product.frameworkName)
+                try? await cacheSystem.cacheFramework(at: outputPath, product: product)
+
+                if case .prepareDependencies = mode {
+                    try await generateVersionFile(for: product, using: cacheSystem)
+                }
+            }
+        }
+    }
+
+    private func buildXCFrameworkIfNeeded(
+        _ product: BuildProduct,
+        mode: Runner.Mode,
+        outputDir: URL,
+        cacheSystem: CacheSystem,
+        compiler: Compiler<some Executor>
+    ) async throws {
+        let frameworkName = product.frameworkName
+        let outputPath = outputDir.appendingPathComponent(product.frameworkName)
+        let exists = fileSystem.exists(outputPath)
+
+        let needToBuild: Bool
+        if exists && isCacheEnabled {
+            let isValidCache = await cacheSystem.existsValidCache(product: product)
+            if isValidCache {
+                logger.info("✅ Valid \(product.target.name).xcframework is exists. Skip building.", metadata: .color(.green))
+                return
+            } else {
+                logger.warning("⚠️ Existing \(frameworkName) is outdated.", metadata: .color(.yellow))
+                logger.info("💥 Delete \(frameworkName)", metadata: .color(.red))
+                try fileSystem.removeFileTree(at: outputPath)
+                let restored = await cacheSystem.restoreCacheIfPossible(product: product)
+                needToBuild = !restored
+                if restored {
+                    logger.info("✅ Restore \(frameworkName) from cache storage", metadata: .color(.green))
+                }
```

Comment:
Why can we restore cache even the cache is invalid?

Reply:
The invalid caches exist on an output dir ( e.g. `$PROJECT_DIR/XCFrameworks`)

On the other hand, Scipio tries to restore from other cache dirs (e.g. `~/Library/Caches`).

If XCFrameworks in output dirs are invalid, then they will be restored from cache dir.
So I think this implementation is correct.